### ISSUE TITLE
The plotting window does not respond when plotting on Windows with the Tk backend

### ISSUE THREAD

**Issue by hroest** [spyder-ide/spyder#9235]
## Problem Description

When trying to plot with the Tk backend on Windows, the plotting window stops responding. Spyder still works, but you have to kill the kernel and start a new one to resume work. Also, no plots are shown. Note, this does not happen with any of the other backends.

### What steps reproduce the problem?

1. Start Spyder
2. Change backend to Tk
3. Restart Spyder
4. Execute

```
import pylab
pylab.plot( [1], [1] )
pylab.show()
```

### What is the expected output? What do you see instead?

Expected: a plot

Instead: The plotting device stops responding and the kernel dies:

![Screenshot from 2019-04-29 09-59-55](https://user-images.githubusercontent.com/464014/56906036-7813fa00-6a6f-11e9-879f-b9dfaa7212ab.png)

### Paste Traceback/Error Below (if applicable)

no traceback

## Versions

* Spyder version: 3.3.5
* Python version: 3.7.3
* Qt version: Qt 5.12.2 
* PyQt version: PyQt5 5.12.1
* Operating System name/version: Windows 10 

### Dependencies

```
IPython >=4.0     :  7.4.0 (OK)
cython >=0.21     :  0.29.6 (OK)
jedi >=0.9.0      :  0.13.3 (OK)
matplotlib >=2.0.0:  3.0.3 (OK)
nbconvert >=4.0   :  5.4.1 (OK)
numpy >=1.7       :  1.16.2 (OK)
pandas >=0.13.1   :  0.24.2 (OK)
pycodestyle >=2.3 :  2.5.0 (OK)
pyflakes >=0.6.0  :  2.1.1 (OK)
pygments >=2.0    :  2.3.1 (OK)
pylint >=0.25     :  2.3.1 (OK)
qtconsole >=4.2.0 :  4.4.3 (OK)
rope >=0.9.4      :  0.12.0 (OK)
sphinx >=0.6.6    :  1.8.5 (OK)
sympy >=0.7.3     :  1.3 (OK)
```

**Comment by ccordoba12** [comment]
@CAM-Gerlach, @dalthviz, could you confirm @hroest problem?

**Comment by dalthviz** [comment]
@ccordoba12 I tried to reproduce this but I could not (I used a conda env with `Python 3.6.8`, `Spyder v3.3.4` and `Spyder-kernels v0.4.4`). Although, the behaviour I experienced is that the graph is not shown at all if the inline backend is `Tkinter` (no window has opened while testing).

**Comment by ccordoba12** [comment]
@dalthviz, that's even worst.

@hroest, I think this is not a problem with Spyder but with the Matplotlib packages in Anaconda.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Sorry for the delay.

I get the exact behavior reported by the OP, with a trivial plot (``plt.plot([1, 2, 3])``) and the ``tkinter`` backend. Plotting window opens but is blank, immediately stops responding, I get the OS-level "Python is not responding" message, and kernel dies upon closing it. The error I get:

```python-traceback
ERROR:tornado.application:Exception in callback functools.partial(.null_wrapper at 0x0000000008627950>)
Traceback (most recent call last):
File "C:\Miniconda3\envs\glml\lib\site‑packages\tornado\ioloop.py", line 758, in _run_callback
ret = callback()
File "C:\Miniconda3\envs\glml\lib\site‑packages\tornado\stack_context.py", line 300, in null_wrapper
return fn(*args, **kwargs)
File "C:\Miniconda3\envs\glml\lib\site‑packages\ipykernel\kernelbase.py", line 306, in advance_eventloop
eventloop(self)
File "C:\Miniconda3\envs\glml\lib\site‑packages\ipykernel\eventloops.py", line 238, in loop_tk
app.tk.createfilehandler(stream.getsockopt(zmq.FD), READABLE, notifier)
AttributeError: '_tkinter.tkapp' object has no attribute 'createfilehandler'
```

This also happens running Spyder 4 ``master`` with a kernel in two different envs, one same as Spyder's and one different, with two different Python versions (3.6.8 and 3.7.2), on Windows 8.1 x64 with Miniconda 2018.12 x64 and PyQt 5.9.2/Qt 5.9.7. Full env details:

Spyder env (there's some extra stuff in it right now so packages affected by the environment ``PATH`` but I reported can load their DLLs) :

<details>

```
# packages in environment at C:\Miniconda3\envs\spyder-4:
#
# Name                    Version                   Build  Channel
alabaster                 0.7.12                   py37_0
asn1crypto                0.24.0                   py37_0
astroid                   2.1.0                    py37_0
atomicwrites              1.3.0                    py37_1
attrs                     18.2.0           py37h28b3542_0
autopep8                  1.4.3                    py37_0
babel                     2.6.0                    py37_0
backcall                  0.1.0                    py37_0
blas                      1.0                         mkl
bleach                    3.1.0                    py37_0
bzip2                     1.0.6                hfa6e2cd_5
ca-certificates           2019.1.23                     0
cartopy                   0.17.0           py37h5ae9855_1
certifi                   2019.3.9                 py37_0
cffi                      1.11.5           py37h74b6da3_1
cftime                    1.0.3.4          py37h8c2d366_0
chardet                   3.0.4                    py37_1
click                     7.0                      py37_0
click-plugins             1.1.1                      py_0
cligj                     0.5.0                    py37_0
cloudpickle               0.7.0                      py_0
colorama                  0.4.1                    py37_0
coverage                  4.5.2            py37he774522_0
cryptography              2.5              py37h7a1dbc1_0
curl                      7.64.1               h2a8f88b_0
cycler                    0.10.0                   py37_0
cython                    0.29.4           py37ha925a31_0
decorator                 4.3.2                    py37_0
defusedxml                0.5.0                    py37_1
descartes                 1.1.0                    py37_0
docutils                  0.14                     py37_0
entrypoints               0.3                      py37_0
expat                     2.2.5                he025d50_0
fastcache                 1.0.2            py37hfa6e2cd_2
fiona                     1.8.4            py37h22081e2_0
flaky                     3.5.3                    py37_0
freetype                  2.9.1                ha9979f8_1
freexl                    1.0.5                hfa6e2cd_0
future                    0.17.1                   py37_0
gdal                      2.3.3            py37hdf43c64_0
geopandas                 0.4.1                      py_0
geos                      3.7.1                h33f27b4_0
hdf4                      4.2.13               h712560f_2
hdf5                      1.10.4               h7ebc959_0
icc_rt                    2019.0.0             h0cc432a_1
icu                       58.2                 ha66f8fd_1
idna                      2.8                      py37_0
imagesize                 1.1.0                    py37_0
intel-openmp              2019.1                      144
ipykernel                 5.1.0            py37h39e3cac_0
ipython                   7.2.0            py37h39e3cac_0
ipython_genutils          0.2.0                    py37_0
isort                     4.3.4                    py37_0
jedi                      0.13.2                   py37_0
jinja2                    2.10                     py37_0
jpeg                      9b                   hb83a4c4_2
jsonschema                2.6.0                    py37_0
jupyter_client            5.2.4                    py37_0
jupyter_core              4.4.0                    py37_0
kealib                    1.4.7                h07cbb95_6
keyring                   17.1.1                   py37_0
kiwisolver                1.0.1            py37h6538335_0
krb5                      1.16.1               hc04afaa_7
lazy-object-proxy         1.3.1            py37hfa6e2cd_2
libboost                  1.67.0               hd9e427e_4
libcurl                   7.64.1               h2a8f88b_0
libgdal                   2.3.3                h10f50ba_0
libiconv                  1.15                 h1df5818_7
libkml                    1.3.0                he5f2a48_4
libnetcdf                 4.6.1                h411e497_2
libpng                    1.6.36               h2a8f88b_0
libpq                     11.2                 h3235a2c_0
libsodium                 1.0.16               h9d3ae62_0
libspatialindex           1.8.5                h6538335_2
libspatialite             4.3.0a              hc36aec2_19
libssh2                   1.8.2                h7a1dbc1_0
libtiff                   4.0.10               hb898794_2
libxml2                   2.9.9                h464c3ec_0
libxslt                   1.1.33               h579f668_0
lxml                      4.3.3            py37h1350720_0
mapclassify               2.0.1                      py_0
markupsafe                1.1.0            py37he774522_0
matplotlib                3.0.2            py37hc8f65d3_0
mccabe                    0.6.1                    py37_1
mistune                   0.8.4            py37he774522_0
mkl                       2019.1                      144
mkl_fft                   1.0.10           py37h14836fe_0
mkl_random                1.0.2            py37h343c172_0
mock                      2.0.0                    py37_0
more-itertools            5.0.0                    py37_0
mpmath                    1.1.0                    py37_0
munch                     2.3.2                    py37_0
nbconvert                 5.4.0                    py37_1
nbformat                  4.4.0                    py37_0
netcdf4                   1.4.2            py37h812ae01_0
numpy                     1.15.4           py37h19fb1c0_0
numpy-base                1.15.4           py37hc3f5095_0
numpydoc                  0.8.0                    py37_0
olefile                   0.46                     py37_0
openssl                   1.1.1b               he774522_1
owslib                    0.17.1                     py_0
packaging                 19.0                     py37_0
pandas                    0.24.2           py37ha925a31_0
pandoc                    1.19.2.1             hb2460c7_1
pandocfilters             1.4.2                    py37_1
parso                     0.3.2                    py37_0
pbr                       5.1.1                    py37_0
pcre                      8.43                 ha925a31_0
pickleshare               0.7.5                    py37_0
pillow                    5.4.1            py37hdc69c19_0
pip                       19.0.1                   py37_0
pluggy                    0.8.1                    py37_0
proj4                     5.2.0                ha925a31_1
prompt_toolkit            2.0.8                      py_0
psutil                    5.5.0            py37he774522_0
psycopg2                  2.7.6.1          py37h7a1dbc1_0
py                        1.7.0                    py37_0
pycodestyle               2.5.0                    py37_0
pycparser                 2.19                     py37_0
pydocstyle                3.0.0                    py37_0
pyepsg                    0.4.0                    py37_0
pyflakes                  2.1.0                    py37_0
pygments                  2.3.1                    py37_0
pykdtree                  1.3.1            py37h8c2d366_2
pylint                    2.2.2                    py37_0
pyopenssl                 19.0.0                   py37_0
pyparsing                 2.3.1                    py37_0
pyproj                    1.9.6            py37h6782396_0
pyqt                      5.9.2            py37h6538335_2
pyshp                     2.1.0                      py_0
pysocks                   1.6.8                    py37_0
pytest                    4.0.2                    py37_0
pytest-cov                2.6.1                    py37_0
pytest-mock               1.10.0                   py37_0
pytest-qt                 3.2.1                      py_0    spyder-ide
python                    3.7.2                h8c8aaf0_2
python-dateutil           2.7.5                    py37_0
python-jsonrpc-server     0.1.2                      py_0    spyder-ide
python-language-server    0.21.5                py37_1000    spyder-ide
pytz                      2018.9                   py37_0
pywin32                   223              py37hfa6e2cd_1
pyzmq                     17.1.2           py37ha925a31_2
qdarkstyle                2.6.5                      py_0    spyder-ide
qt                        5.9.7            vc14h73c81de_0
qtawesome                 0.5.7                    py37_1
qtconsole                 4.4.3                    py37_0
qtpy                      1.7.0                    py37_1
requests                  2.21.0                   py37_0
rope                      0.11.0                   py37_0
rtree                     0.8.3                    py37_0
scipy                     1.2.0            py37h29ff71c_0
setuptools                40.7.3                   py37_0
shapely                   1.6.4            py37h222a598_0
sip                       4.19.8           py37h6538335_0
six                       1.12.0                   py37_0
snowballstemmer           1.2.1                    py37_0
sphinx                    1.8.4                    py37_0
sphinxcontrib             1.0                      py37_1
sphinxcontrib-websupport  1.1.0                    py37_1
spyder-kernels            1.3.2                    py37_0
sqlalchemy                1.3.3            py37he774522_0
sqlite                    3.26.0               he774522_0
sympy                     1.3                      py37_0
testpath                  0.4.2                    py37_0
tk                        8.6.8                hfa6e2cd_0
tornado                   5.1.1            py37hfa6e2cd_0
traitlets                 4.3.2                    py37_0
urllib3                   1.24.1                   py37_0
vc                        14.1                 h0510ff6_4
vs2015_runtime            14.15.26706          h3a45250_0
wcwidth                   0.1.7                    py37_0
webencodings              0.5.1                    py37_1
wheel                     0.32.3                   py37_0
win_inet_pton             1.0.1                    py37_1
wincertstore              0.2                      py37_0
wrapt                     1.11.1           py37he774522_0
xarray                    0.12.1                     py_0
xerces-c                  3.2.2                ha925a31_0
xz                        5.2.4                h2fa13f4_4
yapf                      0.25.0                   py37_0
zeromq                    4.3.1                h33f27b4_3
zlib                      1.2.11               h62dcd97_3
zstd                      1.3.7                h508b16e_0
```

</details>

Other env:

<details>

```
# packages in environment at C:\Miniconda3\envs\glml:
#
# Name                    Version                   Build  Channel
asn1crypto                0.24.0                   py36_0
attrs                     19.1.0                   py36_1
backcall                  0.1.0                    py36_0
blas                      1.0                         mkl
blosc                     1.15.0               h7bd577a_0
bzip2                     1.0.6                hfa6e2cd_5
ca-certificates           2019.1.23                     0
cachetools                2.1.0                      py_0    conda-forge
cartopy                   0.17.0           py36h5ae9855_1
certifi                   2019.3.9                 py36_0
cffi                      1.12.2           py36h7a1dbc1_1
cftime                    1.0.3.4          py36h8c2d366_0
chardet                   3.0.4                    py36_1
click                     7.0                      py36_0
click-plugins             1.1.1                      py_0
cligj                     0.5.0                    py36_0
cloudpickle               0.7.0                      py_0
colorama                  0.4.1                    py36_0
cryptography              2.6.1            py36h7a1dbc1_0
curl                      7.64.0               h2a8f88b_2
cycler                    0.10.0           py36h009560c_0
decorator                 4.3.2                    py36_0
descartes                 1.1.0                    py36_0
expat                     2.2.5                he025d50_0
fiona                     1.8.4            py36h22081e2_0
freetype                  2.9.1                ha9979f8_1
freexl                    1.0.5                hfa6e2cd_0
gdal                      2.3.3            py36hdf43c64_0
geopandas                 0.4.1                      py_0
geos                      3.7.1                h33f27b4_0
glmtools                  0.1.dev0                  <pip>
google-api-core           1.10.0           py36h21ff451_0    conda-forge
google-auth               1.6.3                      py_0    conda-forge
google-cloud-core         0.29.1                     py_0    conda-forge
google-cloud-storage      1.15.0                     py_0    conda-forge
google-resumable-media    0.3.2                      py_0    conda-forge
googleapis-common-protos  1.5.9            py36h21ff451_2    conda-forge
h5py                      2.9.0            py36h5e291fa_0
hdf4                      4.2.13               h712560f_2
hdf5                      1.10.4               h7ebc959_0
icc_rt                    2019.0.0             h0cc432a_1
icu                       58.2                 ha66f8fd_1
idna                      2.8                      py36_0
intel-openmp              2019.1                      144
ipykernel                 5.1.0            py36h39e3cac_0
ipython                   7.2.0            py36h39e3cac_0
ipython_genutils          0.2.0            py36h3c5d0ee_0
jedi                      0.13.2                   py36_0
joblib                    0.13.1                   py36_0
jpeg                      9b                   hb83a4c4_2
jupyter_client            5.2.4                    py36_0
jupyter_core              4.4.0                    py36_0
kealib                    1.4.7                h07cbb95_6
kiwisolver                1.0.1            py36h6538335_0
krb5                      1.16.1               hc04afaa_7
libboost                  1.67.0               hd9e427e_4
libcurl                   7.64.0               h2a8f88b_2
libgdal                   2.3.3                h10f50ba_0
libiconv                  1.15                 h1df5818_7
libkml                    1.3.0                he5f2a48_4
libnetcdf                 4.6.1                h411e497_2
libpng                    1.6.36               h2a8f88b_0
libpq                     11.2                 h3235a2c_0
libprotobuf               3.6.1                h7bd577a_0
libsodium                 1.0.16               h9d3ae62_0
libspatialindex           1.8.5                h6538335_2
libspatialite             4.3.0a              hc36aec2_19
libssh2                   1.8.0                h7a1dbc1_4
libtiff                   4.0.10               hb898794_2
libxml2                   2.9.9                h464c3ec_0
libxslt                   1.1.33               h579f668_0
lmatools                  0.6a0                     <pip>
lxml                      4.3.2            py36h1350720_0
lzo                       2.10                 h6df0209_2
mapclassify               2.0.1                      py_0
matplotlib                3.0.3            py36hc8f65d3_0
mkl                       2019.3                      203
mkl_fft                   1.0.10           py36h14836fe_0
mkl_random                1.0.2            py36h343c172_0
munch                     2.3.2                    py36_0
netcdf4                   1.4.2            py36h812ae01_0
numexpr                   2.6.9            py36hdce8814_0
numpy                     1.16.3           py36h19fb1c0_0
numpy-base                1.16.3           py36hc3f5095_0
olefile                   0.46                     py36_0
openssl                   1.1.1b               he774522_1
owslib                    0.17.1                     py_0
pandas                    0.24.2           py36ha925a31_0
parso                     0.3.2                    py36_0
patsy                     0.5.1                    py36_0
pcre                      8.43                 ha925a31_0
pickleshare               0.7.5                    py36_0
pillow                    5.4.1            py36hdc69c19_0
pip                       19.0.1                   py36_0
proj4                     5.2.0                ha925a31_1
prompt_toolkit            2.0.8                      py_0
protobuf                  3.6.1            py36h33f27b4_0
psycopg2                  2.7.6.1          py36h7a1dbc1_0
pyasn1                    0.4.5                      py_0
pyasn1-modules            0.2.4                    py36_0
pycparser                 2.19                     py36_0
pyepsg                    0.4.0                    py36_0
pygments                  2.3.1                    py36_0
pykdtree                  1.3.1            py36h8c2d366_2
pyopenssl                 19.0.0                   py36_0
pyparsing                 2.3.1                    py36_0
pyproj                    1.9.6            py36h6782396_0
pyqt                      5.9.2            py36h6538335_2
pyreadline                2.1                      py36_1
pyshp                     2.0.1                    py36_0
pysocks                   1.6.8                    py36_0
pytables                  3.5.1            py36h1da0976_0
python                    3.6.8                h9f7ef89_1
python-dateutil           2.7.5                    py36_0
pytz                      2018.9                   py36_0
pyzmq                     17.1.2           py36ha925a31_2
qt                        5.9.7            vc14h73c81de_0
requests                  2.21.0                   py36_0
rsa                       3.4.2                      py_1    conda-forge
rtree                     0.8.3                    py36_0
scikit-learn              0.20.2           py36h343c172_0
scipy                     1.2.1            py36h29ff71c_0
setuptools                40.7.3                   py36_0
shapely                   1.6.4            py36h222a598_0
sip                       4.19.8           py36h6538335_0
six                       1.12.0                   py36_0
snappy                    1.1.7                h777316e_3
spyder-kernels            1.3.2                    py36_0
sqlalchemy                1.3.3            py36he774522_0
sqlite                    3.26.0               he774522_0
statsmodels               0.9.0            py36h452e1ab_0
tk                        8.6.8                hfa6e2cd_0
tornado                   5.1.1            py36hfa6e2cd_0
traitlets                 4.3.2            py36h096827d_0
urllib3                   1.24.1                   py36_0
vc                        14.1                 h0510ff6_4
vs2015_runtime            14.15.26706          h3a45250_0
wcwidth                   0.1.7            py36h3d5aa90_0
wheel                     0.32.3                   py36_0
win_inet_pton             1.1.0                    py36_0
wincertstore              0.2              py36h7fe50ca_0
xarray                    0.12.1                     py_0
xerces-c                  3.2.2                ha925a31_0
xz                        5.2.4                h2fa13f4_4
zeromq                    4.3.1                h33f27b4_3
zlib                      1.2.11               h62dcd97_3
zstd                      1.3.7                h508b16e_0
```

</details>

**Comment by goanpeca** [comment]
Related to https://github.com/ipython/ipykernel/issues/358

**Comment by gplonline** [comment]
I am facing exactly the same issue as hroest's. Has there been progress? A solution / work-around?

**Comment by ccordoba12** [comment]
@gplonline, please use another backend instead of Tk. That's the only workaround for now.

**Comment by yesinnk** [comment]
@ccordoba12 so does ipython no longer support event loop integration with Tk?

**Comment by ccordoba12** [comment]
It's a bug that no one has time to fix.

**Comment by yesinnk** [comment]
@ccordoba12 sad, what's the work around other than downgrading the ipykernel version?

**Comment by ccordoba12** [comment]
I really don't know, sorry.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 We could disable the Tk option on Windows in preferences if IPykernel > 4.10 until the bug is solved, but the problem is the version varies by environment so we have no way of knowing a priori if the user is going to open a kernel in an environment that does or doesn't support it. Furthermore, it mean a lag until it works if and when a version is released supporting Tk. Maybe show a warning dialog when Tk is selected explaining the problem?

**Comment by ccordoba12** [comment]
@CAM-Gerlach, I think it's easier to fix the bug ourselves than implementing the workaround you're suggesting.

Would you like to take a look at it?

**Comment by CAM-Gerlach** [comment]
@ccordoba12 I've never worked with Tk/Tkinter, async event looks or on `ipykernel` and I have never used Tk graphics so I don't think I'm the best person for the job, sorry. @yesinnk is there something in particular that only the Tk Matplotlib backend provides and the others do not, that might motivate working on this?

**Comment by yesinnk** [comment]
@CAM-Gerlach I'm trying to integrate the `tk` event loop in the `ipykernel` (jupyter notebook to be exact) and whenever I run `%gui tk` I get `AttributeError: '_tkinter.tkapp' object has no attribute 'createfilehandler'`

**Comment by CAM-Gerlach** [comment]
Yup; as I believe you noticed on the issue, creating file handlers is not supported on Windows due to its different OS interfaces, primitives and abstractions. I'm not really familiar with the mechanics of this and what might be the best workaround; the `ipykernel` devs are likely the best ones to ask (as you did).

**Comment by yesinnk** [comment]
thanks for getting back to me @CAM-Gerlach could you refer me to the `ipykernel` devs / where can I get in touch with them?

**Comment by smichalove** [comment]
Not sure if this is related to your issue, but I was having a similar issue and I found:

If you need to wait for the next sample do not use sleep, use matplotlib.pyplot.pause(interval) https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.pause.html

**Comment by ccordoba12** [comment]
This error is fixed in our latest version (5.3.1), so please update.
